OAK-10341 Tree store

thomasmueller opened the pull request:
Ready for review.

reschke committed:
OAK-10944: oak-auth-ldap: update commons-pool2 dependency to 2.12.0 (#1576)

mbaedke committed:
OAK-10705: oak-standalone: update dependencies (#1411)

Updated dependencies.

thomasmueller committed:
OAK-10341 Tree store

thomasmueller committed:
OAK-10341 Tree store

nit0906 commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/tree/TreeStore.java:
```diff
@@ -0,0 +1,169 @@
+import org.apache.jackrabbit.oak.plugins.memory.EmptyNodeState;
+import org.apache.jackrabbit.oak.spi.state.NodeState;
+
+public class TreeStore implements Iterable<NodeStateEntry>, Closeable {
```
I see that this is draft right now - but we should probably implement the IndexStore interface to make it easier to integrate.

thomasmueller committed:
OAK-10341 Tree store

nit0906 committed:
OAK-10905 | Add a configurable  async checkpoint creator service (#1560)

nit0906 commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/tree/TreeStore.java:
```diff
@@ -166,4 +171,28 @@ public Store getStore() {
+
+    @Override
+    public boolean isIncremental() {
+        return true;
```
This should be false. 

The TreeStore should be the base implementation - that builds the a tree representation of the whole repository, and then we can have something like an IncrementalTreeStore, which builds the diff b/w 2 checkpoints to be appended to the base TreeStore.

Or maybe simply use the diff by IncrementalFlatFileStore and merge that in the base TreeStore.

Refer - 

IndexStore -> FlatFileStore -> IncrementalFlatFileStore

thomasmueller replied:
To update the tree store incrementally, we can use log-structured merges, using Session.checkpoint() / Session.flush(). In theory, this is much more efficient than the IncrementalFlatFileStore, if done correctly. Specially if there are few changes. This is already implemented in the tree store, but needs to be "wired" correctly. 

But I think I will not yet do that just yet, but instead concentrate on getting the basics working correctly (right now, e.g. the cache size is in number of entries instead of memory used... this can lead to OOME, or then in too low memory usage and so bad performance, if we have very unevenly sizes pages -- and we probably have this). Also, I need to port some unit tests.

So I'll set it to "false" currently.

thomasmueller committed:
OAK-10341 Tree store

nit0906 committed:
OAK-10905 | Add license header to  AsyncCheckpointService (#1579)

thomasmueller committed:
OAK-10341 Tree store

reschke committed:
OAK-10848: commons: remove use of slf4j.event.Level in SystemPropertySupplier implementation (#1580)

reschke committed:
OAK-10685: remove unused import of java.io.UnsupportedEncodingException

reschke committed:
OAK-10949: blob-cloud, segment-aws: update aws SDK to 1.12.761 (dependencies reference vulnerable amazon ion-java version) (#1581)

reschke committed:
OAK-10954: Update spotbugs plugin to 4.8.6.2 (#1588)

reschke committed:
OAK-10959: webapp: update Tomcat dependency to 9.0.90 (#1589)

reschke committed:
OAK-10960: blob-cloud, segment: update netty version to 4.1.111 (#1590)

reschke committed:
OAK-10945: Remove usage of Guava Function interface (#1578)

* OAK-10945: Remove usage of Guava Function interface (oak-upgrade)

* OAK-10945: Remove usage of Guava Function interface (oak-store-spi)

* OAK-10945: Remove usage of Guava Function interface (oak-it)

* OAK-10945: Remove usage of Guava Function interface (oak-store-document)

* OAK-10945: Remove usage of Guava Function interface (oak-store-composite)

* OAK-10945: Remove usage of Guava Function interface (oak-segment-tar)

* OAK-10945: Remove usage of Guava Function interface (oak-segment-azure)

* OAK-10945: Remove usage of Guava Function interface (oak-security-spi)

* OAK-10945: Remove usage of Guava Function interface (oak-security-search)

* OAK-10945: Remove usage of Guava Function interface (oak-run-commons)

* OAK-10945: Remove usage of Guava Function interface (oak-run)

* OAK-10945: Remove usage of Guava Function interface (oak-lucene)

* OAK-10945: Remove usage of Guava Function interface (oak-jcr)

* OAK-10945: Remove usage of Guava Function interface (oak-exercise)

* OAK-10945: Remove usage of Guava Function interface (oak-core-spi)

* OAK-10945: Remove usage of Guava Function interface (oak-core)

* OAK-10945: Remove usage of Guava Function interface (oak-commons)

* OAK-10945: Remove usage of Guava Function interface (oak-blob-plugins)

* OAK-10945: Remove usage of Guava Function interface (oak-blob-cloud, oak-blob-cloud-azure)

* OAK-10945: Remove usage of Guava Function interface (oak-upgrade) - cleanup

* OAK-10945: Remove usage of Guava Function interface (oak-store-spi) - cleanup

* OAK-10945: Remove usage of Guava Function interface (oak-store-document) - cleanup

* OAK-10945: Remove usage of Guava Function interface (oak-store-composite) - cleanup

* OAK-10945: Remove usage of Guava Function interface (oak-segment-tar) - cleanup

* OAK-10945: Remove usage of Guava Function interface (oak-lucene) - cleanup

* OAK-10945: Remove usage of Guava Function interface (oak-jcr) - cleanup

reschke committed:
OAK-10962: oak-solr-osgi: update zookeeper dependency to 3.9.2 (#1591)

reschke committed:
Update build.yml to disable Sonar for now

...because of failures, as in https://github.com/apache/jackrabbit-oak/actions/runs/10021673018

nfsantos committed:
OAK-10965 - Make ConsoleIndexingReporter thread safe. (#1592)

mbaedke committed:
OAK-6762: Convert oak-blob to OSGi R7 annotations (#1413)

done

mbaedke committed:
OAK-6773: Convert oak-store-composite to OSGi R7 annotations (#1489)

done

nfsantos committed:
OAK-10951 - Add a new configuration property (#1594)

- "oak.indexer.persistedLinkedList.cacheSize" - sets the cache size of the PersistedLinkedList used to traverse the FFS. This controls the number FFS entries kept in memory.

nfsantos committed:
OAK-10966 - Avoid object creation in PathUtils.isAncestor (#1596)

t-rana committed:
OAK-10964: bump nimbus-jose-jwt dependency to latest (#1593)

reschke committed:
OAK-6773: Convert oak-store-composite to OSGi R7 annotations - fix line ends

ionutzpi committed:
OAK-10803 -- compress/uncompress property, disabled by default (#1526)

Co-authored-by: pirlogea <pirlogea@adobe.com>

stefan-egli committed:
OAK-10974 and OAK-10869 : temporarily disabling flaky tests

nfsantos committed:
OAK-10971 - Add a method to test if a path is a direct ancestor of another: PathUtils.isDirectAncestor()  (#1598)

reschke committed:
OAK-10803: fix NPE when Mongo is unavailable, remove '*' imports (#1601)

nfsantos committed:
OAK-10976 - Avoid unnecessary call to PathUtils.getName in IndexDefinition (#1603)

t-rana committed:
OAK-10904: Close token refresh executor service after access token is no longer needed (#1545)

* OAK-10904: change token refresh executor to use daemon thread

* OAK-10904: add log

* OAK-10904: move storage credential method to concrete class and explicitly calling close to shutdown the executor

* OAK-10904: wrap code for generating storage credentials in try catch

* OAK-10904: minor changes

* OAK-10904: refactor log

nfsantos committed:
OAK-10977 - Cleanup IndexDefinition class (#1604)

- replace Guava usages with JDK equivalents
- fix typos
- improve formatting

dulceanu committed:
OAK-10978 - Skip Azure compaction when there's not enough garbage in the repository (#1606)

nfsantos committed:
OAK-10966 - Indexing job: create optimized version of PersistedLinkedList (#1595)

thomasmueller committed:
Revert "OAK-10966 - Indexing job: create optimized version of PersistedLinkedList (#1595)"

This reverts commit 8a72ef89022e6187bb878400301aab93bd16af88.

thomasmueller committed:
Revert "OAK-10978 - Skip Azure compaction when there's not enough garbage in the repository (#1606)"

This reverts commit 58146387d61e92700bf3f75a1b78d2f87e83070b.

thomasmueller committed:
Revert "OAK-10977 - Cleanup IndexDefinition class (#1604)"

This reverts commit ce5c7df8993d2319828fceaacdcfc2093d511919.

thomasmueller committed:
Revert "OAK-10904: Close token refresh executor service after access token is no longer needed (#1545)"

This reverts commit 632a15b3d0af8c98c65e4a02b40743db94729184.

thomasmueller committed:
Revert "OAK-10976 - Avoid unnecessary call to PathUtils.getName in IndexDefinition (#1603)"

This reverts commit e4119605ebfe8fe7c4d2b02805291e3605af9134.

thomasmueller committed:
Revert "OAK-10803: fix NPE when Mongo is unavailable, remove '*' imports (#1601)"

This reverts commit 5dd63444fac480e104966babbae076afdb618bb0.

thomasmueller committed:
Revert "OAK-10971 - Add a method to test if a path is a direct ancestor of another: PathUtils.isDirectAncestor()  (#1598)"

This reverts commit c416850e3aaca0a6057aa3541531423a34f763a8.

thomasmueller committed:
Revert "OAK-10974 and OAK-10869 : temporarily disabling flaky tests"

This reverts commit aefb990391f51a8ea354a68f9126d4aec959e191.

thomasmueller committed:
Revert "OAK-10803 -- compress/uncompress property, disabled by default (#1526)"

This reverts commit 25792e7c929052a434cc0c9ee117883dc1793b79.

thomasmueller committed:
Revert "OAK-6773: Convert oak-store-composite to OSGi R7 annotations - fix line ends"

This reverts commit 818317c7200914cb27b7c4f791fb996e75a74dc6.

thomasmueller committed:
Revert "OAK-10964: bump nimbus-jose-jwt dependency to latest (#1593)"

This reverts commit 9528fddef84449ea71f6b36dc484316afd421b0d.

thomasmueller committed:
Revert "OAK-10966 - Avoid object creation in PathUtils.isAncestor (#1596)"

This reverts commit b37db4c507b9f4efd0651db8185743c18d90477d.

thomasmueller committed:
Revert "OAK-10951 - Add a new configuration property (#1594)"

This reverts commit ebefe01536172e55c98fad969e30eb448408324c.

thomasmueller committed:
Revert "OAK-6773: Convert oak-store-composite to OSGi R7 annotations (#1489)"

This reverts commit 0521c6376caa5d2c5e8b0322df3840e49aaf46ff.

thomasmueller committed:
Revert "OAK-6762: Convert oak-blob to OSGi R7 annotations (#1413)"

This reverts commit 6c44805441cc92542c13deabb0afc4af68ec6bc3.

thomasmueller committed:
Revert "OAK-10965 - Make ConsoleIndexingReporter thread safe. (#1592)"

This reverts commit 8b05caefb94252c49b62eb2499a705f3d4dd09f8.

thomasmueller committed:
Revert "Update build.yml to disable Sonar for now"

This reverts commit 8db9183b6f2de4d6bfc3ea12cffbdf705a868016.

thomasmueller committed:
Revert "OAK-10962: oak-solr-osgi: update zookeeper dependency to 3.9.2 (#1591)"

This reverts commit 95ed6c46a217c46c075747f2fbc1fed227c5504a.

thomasmueller committed:
Revert "OAK-10945: Remove usage of Guava Function interface (#1578)"

This reverts commit 5d56db1224f91992db8fb2978b2a4a482a7ca705.

thomasmueller committed:
Revert "OAK-10960: blob-cloud, segment: update netty version to 4.1.111 (#1590)"

This reverts commit cd35db705fe768ede6991c6a5d25de034081114b.

thomasmueller committed:
Revert "OAK-10959: webapp: update Tomcat dependency to 9.0.90 (#1589)"

This reverts commit 5fb55e05e0ba137024e07d66dc3856864b87eacf.

thomasmueller committed:
Revert "OAK-10954: Update spotbugs plugin to 4.8.6.2 (#1588)"

This reverts commit b098fd8930ec91cd4ea15769c21db8e4ee47864d.

thomasmueller committed:
Revert "OAK-10949: blob-cloud, segment-aws: update aws SDK to 1.12.761 (dependencies reference vulnerable amazon ion-java version) (#1581)"

This reverts commit cafbd29d520c247a74e88314b1f518c95fb74443.

thomasmueller committed:
Revert "OAK-10685: remove unused import of java.io.UnsupportedEncodingException"

This reverts commit 36cfcfa08e66ef2aecc5cec3c08e6dd08db9038e.

thomasmueller committed:
Revert "OAK-10848: commons: remove use of slf4j.event.Level in SystemPropertySupplier implementation (#1580)"

This reverts commit ba530fe46563757474a4ed1740aa3396e5f051bd.

thomasmueller committed:
Revert "OAK-10905 | Add license header to  AsyncCheckpointService (#1579)"

This reverts commit 7f112faca966859e8bf117bcbd4665892f00a330.

thomasmueller committed:
Revert "OAK-10905 | Add a configurable  async checkpoint creator service (#1560)"

This reverts commit f12ccb1716e6677b54ad3f06955cc007479584bf.

thomasmueller committed:
Revert "OAK-10705: oak-standalone: update dependencies (#1411)"

This reverts commit dc10a1206566d27004aea9aa7f9612547d2078cd.

thomasmueller committed:
Revert "OAK-10944: oak-auth-ldap: update commons-pool2 dependency to 2.12.0 (#1576)"

This reverts commit 3d6d009d654fc8061baac1b66502e47f71fa0921.

thomasmueller committed:
Merge branch 'trunk' into OAK-10341b

thomasmueller committed:
Merge branch 'trunk' into OAK-10341b

thomasmueller committed:
OAK-10341 Tree store (bugfix for loggers)

thomasmueller committed:
OAK-10341 Tree store (PipelinedTreeStoreStrategy.java)

thomasmueller committed:
OAK-10341 Tree store (tests)

thomasmueller committed:
Merge branch 'trunk' into OAK-10341b

thomasmueller committed:
OAK-10341 Tree store (use less memory)

thomasmueller committed:
OAK-10341 Tree store (fix memory cache calculation)

thomasmueller committed:
OAK-10341 Tree store (blob prefetch)

thomasmueller committed:
OAK-10341 Tree store (blob prefetch)

thomasmueller committed:
OAK-10341 Tree store (blob prefetch)

thomasmueller committed:
OAK-10341 Tree store (node prefetch)

thomasmueller committed:
Merge trunk

thomasmueller committed:
OAK-10341 Tree store (incremental)

thomasmueller committed:
OAK-10341 Tree store (pack files)

thomasmueller committed:
Merge branch 'trunk' into OAK-10341b

thomasmueller committed:
OAK-10341 Tree store (pack files)

thomasmueller committed:
OAK-10341 Tree store (tests)

thomasmueller committed:
OAK-10341 Tree store (incremental)

thomasmueller committed:
Merge branch 'trunk' into OAK-10341b

thomasmueller committed:
OAK-10341 Tree store (compress)

thomasmueller committed:
OAK-10341 Tree store (traverse included paths)

thomasmueller committed:
OAK-10341 Tree store

thomasmueller committed:
OAK-10341 Tree store

thomasmueller committed:
OAK-10341 Tree store

thomasmueller committed:
OAK-10341 Tree store

thomasmueller committed:
OAK-10341 Tree store

thomasmueller committed:
Merge branch 'trunk' into OAK-10341b

thomasmueller committed:
OAK-10341 Tree store (use same blob prefetching configuration as for FlatFileStore)

nfsantos commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/DocumentStoreIndexerBase.java:
```diff
@@ -239,6 +239,12 @@ nodeStore, getMongoDocumentStore(), traversalLog))
     }
 
+    public IndexStore buildTreeStore() throws IOException, CommitFailedException {
+        System.setProperty(FlatFileNodeStoreBuilder.OAK_INDEXER_SORT_STRATEGY_TYPE,
```
This will leave this system property behind, even after finishing building the tree store, so it changes global state. Couldn't this cause unintended problems?

thomasmueller replied:
I didn't think about that... I'll try to fix it.

I think it's the same way as the "reuse existing flat file store" works btw. we might also revisit these other usages.

thomasmueller replied:
Fixed here.

nfsantos commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileNodeStoreBuilder.java:
```diff
@@ -19,17 +19,32 @@
-import com.mongodb.MongoClientURI;
-import com.mongodb.client.MongoDatabase;
-import org.apache.commons.lang3.StringUtils;
+import static java.util.Collections.unmodifiableSet;
+import static org.apache.jackrabbit.guava.common.base.Preconditions.checkState;
+import static org.apache.jackrabbit.oak.index.indexer.document.indexstore.IndexStoreUtils.OAK_INDEXER_USE_LZ4;
+import static org.apache.jackrabbit.oak.index.indexer.document.indexstore.IndexStoreUtils.OAK_INDEXER_USE_ZIP;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
```
The order of imports is different from the one of InteliJ. We do not have an agreed formatting style, but since most of us are using InteliJ and the file is already with the InteliJ ordering, I'd prefer we leave it the current order.

thomasmueller replied:
I think everyone uses a slightly different setting... but yes it makes sense to avoid such diffs.

This case should be resolved now.

nfsantos commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileNodeStoreBuilder.java:
```diff
@@ -283,7 +324,7 @@ public List<IndexStore> buildList(IndexHelper indexHelper, IndexerSupport indexe
         // Check system property defined path
         String sortedFilePath = System.getProperty(OAK_INDEXER_SORTED_FILE_PATH);
-        if (StringUtils.isNotBlank(sortedFilePath)) {
+        if (sortedFilePath != null && !sortedFilePath.isEmpty()) {
```
This is not equivalent. If the string contains only whitespace, `StringUtils::isNotBlank` returns false, while this test here returns true, as `#isEmpty` tests that the String contains no characters.

thomasmueller replied:
I'll revert this... I couldn't compile this part of the code due to a problem with the dependency.

(I don't think the usage of "StringUtils.isNotBlank" is intentional btw.)

thomasmueller replied:
Fixed.

nfsantos commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedTransformTask.java:
```diff
@@ -160,7 +160,7 @@ public Result call() throws Exception {
                         statistics.incrementMongoDocumentsTraversed();
                         mongoObjectsProcessed++;
-                        if (mongoObjectsProcessed % 50000 == 0) {
+                        if (mongoObjectsProcessed % 200_000 == 0) {
```
Why the change? I don't find that this log line is printed too often. But adding a thousand separator to the `50000` would be nice.

thomasmueller replied:
In my view we log way too much in general... but OK, I'll change it back.

thomasmueller replied:
Fixed.

nfsantos reviewed (approved):
I have only looked at the parts that interact with existing code, did not review the TreeStore or IndexStore logic.

thomasmueller committed:
OAK-10341 Tree store (javadocs)

thomasmueller committed:
OAK-10341 Tree store (code review)

steffenvan commented:
Just my opinion, but I see no comments that address the logic or readability of the TreeStore. Wouldn't it have made sense to have some reviews on that before merging?

thomasmueller commented:
> I see no comments that address the logic or readability of the TreeStore

Just documenting the tree store (alone) wont resolve that problem. I think we are missing high-level documentation of the implementation for many many features. Including the blob store, the segment store, the document store, oak-core, oak-jcr, etc. I think we should discuss what implementation level-documentation is needed _in general_. I view that as an independent problem.
